pkg/registry/batch/job/strategy.go

```diff
@@ -61,10 +61,9 @@ func (jobStrategy) DefaultGarbageCollectionPolicy(ctx context.Context) rest.Garb
 	}
 	switch groupVersion {
 	case batchv1.SchemeGroupVersion:
-		// for back compatibility
-		return rest.OrphanDependents
```

Comment:
This is reversing the logic and breaking the behavior of delete requests to the v1 Job API this was specifically set up to preserve. Is that intentional?

Reply:
I think, we're leaving the job pods orphaned always to preserve backwards compatibility for the past few releases. Not sure, if it'd make sense to do that now considering we have supported `orphaning job pods` for many releases since the original switch happened. For other resources, there were versions like betav1, alpha2 etc but by the time the change got introduced to job object, it was already v1 and we don't have any versions after that? So, I am wondering if we should support orphaning pods forever in v1 api or introduce a featuregate and phase this behavior because it seems like a bad pattern to orphan job pods always.

Reply:
@liggitt yeah, that's what I concluded when I was talking with @ravisantoshgudimetla about it. His proposal was that we should try to change it though to put it in-line with the rest of controllers, but with the usual process applied. Iow. we'd open a KEP with a feature flag and slowly switch this to `rest.DeleteDependents` over the usual period of 3 releases, wdyt?

Reply:
Thank you for clarifying it @soltysh - Ya, the plan is to phase the behavior over releases. This is to ensure that job controller's latest api is on par with rest of the controllers.

Reply:
ping @liggitt : Can you give us your feedback? Without this feature, we're noticing that the job's pods are always left orphaned for our customers and for prow instance we use internally at Red Hat.

Reply:
I have very mixed feelings about this.

On the one hand, I have a hard time imagining someone preferring the current orphan behavior if given a choice, and it's clear that the majority of users that encounter this are negatively impacted by the current behavior.

On the other hand, I can absolutely imagine automated tools that would be broken by this change, which were not impacted by the leftover pods. For example, a tool that:
1. creates a namespace
2. creates a job
3. waits for the job to complete
4. deletes the job
5. gathers results from the child pods
6. deletes the namespace

I would probably start by trying to surface warnings to API clients that delete jobs without specifying orphanDependents or propagationPolicy encouraging them to explicitly indicate what they want (if something wants to intentionally orphan, it should set `propagationPolicy=Orphan`, if something wants to clean up children, it should set `propagationPolicy=Background`.

Does anything else eventually clean up the completed job pods, like pod GC?

Reply:
cc @smarterclayton for thoughts on v1 compatibility vs danger of the current behavior

Reply:
> I would probably start by trying to surface warnings to API clients that delete jobs without specifying orphanDependents or propagationPolicy encouraging them to explicitly indicate what they want (if something wants to intentionally orphan, it should set propagationPolicy=Orphan

That seems reasonable.

> Does anything else eventually clean up the completed job pods, like pod GC? 

PodGC does clean up eventually.

Reply:
Since PodGC is non deterministic (if you have 10k pods you'll get cleaned up reasonably fast, if you have less than 10k pods you'll get cleaned up very far in the future), we might be able to look at something on the job controller that cleans up pods after a few minutes, EVEN if it's orphan background.  And then explicitly require a field on "minimum time to preserve pods that are completed" with a low default.  Then after that's been out for a while, reassess (with metrics! and warnings!) and potentially suggest a change in behavior.

Reply:
> Since PodGC is non deterministic 

Yeah and 12.5k pods should be lying around in etcd by default to start gc'ing

> we might be able to look at something on the job controller that cleans up pods after a few minutes, EVEN if it's orphan background. 

That's a good idea for long term. We already have TTL controller which is supposed to delete the jobs and it was mentioned that we would like to[ delete the pods](https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/592-ttl-after-finish/README.md#future-work) associated as well in future. We can see for the presence of `job-name` and  absence of `ownerRef` on the pod's metadata to see if it is orphaned and delete it for now. In future, we can see how many pods are lingering like this in the cluster via a metric and see if we can introduce the change in the API v1 directly or in a new version.


Short term as suggested, we will proceed with warning.

Reply:
as an aside, it looks like `kubectl delete`, cronjob controller, and garbagecollection controller default to sending `propagationPolicy="Background"` so all of the in-tree job deleters do not cause pod orphaning

Reply:
Yeah. I noticed that too. Here is [WIP PR](https://github.com/kubernetes/kubernetes/pull/104080), let me know if that looks in right direction.

Reply:
@liggitt - I noticed that TTLController is using [Dependent](https://github.com/kubernetes/kubernetes/blob/58617e161933277fe362599ce380bebf08cbdacb/pkg/controller/ttlafterfinished/ttlafterfinished_controller.go#L236) deletion policy explicitly. I am thinking of adding a warning when job gets created if there is no TTL setup for it.

Reply:
only if there's no controller ownerReference, perhaps?

Reply:
Ya, to be clear, no backing controller like cronjob etc, right?